src/amqproxy/upstream.cr

```diff
@@ -154,66 +178,70 @@ module AMQProxy
-          authentication_failure_close: true,
-          per_consumer_qos:             true,
-          "basic.nack":                 true,
-          direct_reply_to:              true,
-          publisher_confirms:           true,
-          consumer_cancel_notify:       true,
-        },
-      })
-      start_ok = AMQ::Protocol::Frame::Connection::StartOk.new(response: "\u0000#{user}\u0000#{password}",
-        client_properties: props, mechanism: "PLAIN", locale: "en_US")
+      AMQ::Protocol::Frame.from_io(@socket).as(AMQ::Protocol::Frame::Connection::Start)
+
+      response = "\u0000#{credentials.user}\u0000#{credentials.password}"
+      start_ok = AMQ::Protocol::Frame::Connection::StartOk.new(response: response, client_properties: ClientProperties, mechanism: "PLAIN", locale: "en_US")
       start_ok.to_io @socket, IO::ByteFormat::NetworkEndian
       @socket.flush
 
-      tune = AMQ::Protocol::Frame.from_io(@socket, IO::ByteFormat::NetworkEndian) { |f| f.as(AMQ::Protocol::Frame::Connection::Tune) }
-      tune_ok = AMQ::Protocol::Frame::Connection::TuneOk.new(tune.channel_max, tune.frame_max, tune.heartbeat)
+      tune = AMQ::Protocol::Frame.from_io(@socket).as(AMQ::Protocol::Frame::Connection::Tune)
+      channel_max = tune.channel_max.zero? ? UInt16::MAX : tune.channel_max
+      tune_ok = AMQ::Protocol::Frame::Connection::TuneOk.new(channel_max, tune.frame_max, tune.heartbeat)
       tune_ok.to_io @socket, IO::ByteFormat::NetworkEndian
       @socket.flush
 
-      open = AMQ::Protocol::Frame::Connection::Open.new(vhost: vhost)
+      open = AMQ::Protocol::Frame::Connection::Open.new(vhost: credentials.vhost)
       open.to_io @socket, IO::ByteFormat::NetworkEndian
       @socket.flush
 
-      AMQ::Protocol::Frame.from_io(@socket, IO::ByteFormat::NetworkEndian) do |f|
-        case f
-        when AMQ::Protocol::Frame::Connection::Close
-          close_ok = AMQ::Protocol::Frame::Connection::CloseOk.new
-          close_ok.to_io @socket, IO::ByteFormat::NetworkEndian
-          @socket.flush
-          @socket.close
-          raise AccessError.new f.reply_text
-        when AMQ::Protocol::Frame::Connection::OpenOk
-          true
-        end
+      case f = AMQ::Protocol::Frame.from_io(@socket, IO::ByteFormat::NetworkEndian)
+      when AMQ::Protocol::Frame::Connection::OpenOk
+      when AMQ::Protocol::Frame::Connection::Close
+        close_ok = AMQ::Protocol::Frame::Connection::CloseOk.new
+        close_ok.to_io @socket, IO::ByteFormat::NetworkEndian
+        @socket.flush
+        @socket.close
+        raise AccessError.new f.reply_text
+      else
+        raise "Unexpeceted frame on connection to upstream: #{f}"
       end
+      channel_max
     rescue ex : AccessError
       raise ex
     rescue ex
       @socket.close
       raise Error.new ex.message, cause: ex
     end
 
+    ClientProperties = AMQ::Protocol::Table.new({
+      connection_name: "AMQProxy #{VERSION}",
```

Comment:
I've always wanted this to show which client was "behind" the proxy too, so it would say AMQProxy $version + AMQPLib / PHP 2.12.2

Reply:
Not possible, now each client only gets a channel. But even before it was impossible, because the upstream connections were reused between different downstream clients. The connection information can only be supplied when the connection is made, can't be updated after.